src/Entities/Entity.cpp

```diff
@@ -1662,7 +1662,9 @@ bool cEntity::MoveToWorld(const AString & a_WorldName, bool a_ShouldSendRespawn)
 void cEntity::SetSwimState(cChunk & a_Chunk)
 {
 	int RelY = FloorC(GetPosY() + 0.1);
-	if ((RelY < 0) || (RelY >= cChunkDef::Height - 1))
+	int HeadRelY = CeilC(GetPosY() + GetHeight()) - 1;
+	ASSERT(RelY <= HeadRelY);
+	if ((RelY < 0) || (HeadRelY >= cChunkDef::Height - 1))
```

Comment:
What should happen if an entity's feet are in water but its head is out of the world (and vice versa)?

Reply:
It doesn't matter?

Reply:
If I read the code correctly (haven't tested it), IsSwimming would be false if the head is out of the world but the feet are in water, which means that if an entity were burning then walking in water at the world height would not put them out.

I think the flipped bug is also true, but it's hard to see how that situation would occur (since you'd just fall out of the water).

Reply:
That would be true, however it really isn't an important bug; fixing a crash is more important than walking slightly slower at the top of the world.

Reply:
After this is merged, that bug can be fixed in due course.

Reply:
Fair.